All documentation for the workshop is best viewed from the github.io pages

[Documentation](https://ucdavis-bioinformatics-training.github.io/2019_September_UCD_Microbial_Community_Analysis_Workshop/)

Bioinformatics: Microbial Community Analysis (amplicon community profiling) @ UC Davis
Sept. 3, 2019, 9 a.m. - Sept. 6, 2019, 4:30 p.m.
Organizer - Bioinformatics Core
Contact - UC Davis Bioinformatics Core, training.bioinformatics@ucdavis.edu

High-throughput sequencing (HTS) of entire microbial communities has enabled rapid advancement in our understanding of the composition and functional content of their involvment in climate change, environmental pollution, human health, biotechnology, etc. Using these data we are able to achieve the most complete picture of the taxonomic (i.e., what organisms are there) and functional (i.e., what are those organisms are doing) composition of microbial communities, making it possible to perform investigations that include organisms that were previously intractable with culturing-based methods. With HTS, all organisms contained in an environmental sample are sequenced in a culture-independent manner, using either amplicon based methods to investigate the taxonomic content or whole-genome/transcriptome shotgun-based methods to investigate both the taxonomic and functional content of the sampled community.

**This workshop will end at Noon on Friday!**

This workshop covers amplicon-based microbial analysis and methodology using high throughput sequencing technologies, and ancillary topics, including:

* Differences between community analysis with amplicons versus shotgun metagenomics and metatranscriptomics
* Experimental design
* Sequencing library preparation techniques and considerations (lecture only)
* Discussion of sequencing technologies
* Hardware and software considerations
* Review of command line operations
* Sequence data quality assurance and expectations
* Processing of amplicon reads (using the python application dbcAmplicons)
* Analysis of community structure (using R, phyloseq and vegan)
* Post analysis and visualization (using R)

Not all topics regarding microbial community analysis will be covered nor all potential software resources (Qiime or Mothur will not be covered) as there are just too many. We will present a rich collection of lectures and hands-on sessions, covering both theory and tools. All analysis will be performed using the Linux command line and R/RStudio. Exercises will offer you the chance to explore software and protocols for high throughput microbial community data analysis. Command-line exercises will be run on the Genome Center Compute Cluster, which will be available to you for a week after the workshop, so you can continue to practice these skills.  We will also hold a follow-up Q&A session a week or two after the workshop.

### Industry Sponsors

|:------|:------|------|
| Tuesday | <img src="base_figures/10x.png" alt="10x genomics" width="150px"/> |    |  
| Wednesday |  <img src="base_figures/illumina.png" alt="illumina" width="150px"/> |    |  

### FAQ

What will I learn? … This workshop will prepare you to work independently on basic analyses, using your own computing resources or publicly available on-demand computation.  Documentation from previous workshop can be viewed at http://bioinformatics.ucdavis.edu/training/documentation/

Who should attend? … Prior workshop participants have included faculty, post docs, grad students, staff, and industry researchers.

What are the prerequisites? … It is recommended that participants have basic familiarity with the Linux/Unix (or Mac) command line, as well as familiarity with general biological concepts and an enthusiasm for learning about microbial data analysis!

What do I need to bring? … You will need to bring your own laptop to use, have a recent version of Java installed, and an application that will allow you to ssh into a server (e.g. Putty for windows, or iTerm for Apple) as well as the ability to connect to the internet via wifi. Computational exercises will be performed on a remote server. We can help with setup early on in the workshop. If your department does not allow you to install software on the laptop you’ll be bringing, please contact us ahead of time.

How  much does it cost? … The registration fee is $1000 for current students, postdocs, staff and faculty at UC campus' or UC National Labs (LBNL, LANL, LLNL), $1200 for other academic, government or non-profit participants and $2000 for other participants. Registration fees include light breakfast, lunch, and snacks, but do not include dinner, lodging or parking fees.

Where is the workshop? … It will be held on the UC Davis campus and will run from 9:00am to 4:30pm on the dates indicated.  **The workshop will end at 12:00PM on Friday** Attendance is limited to 50 participants in order to foster an effective learning environment and ensure sufficient one-on-one attention.

How do I apply? … All registration is “first-come, first-served”. There is no application process.  So, sign up as soon as possible once registration opens to ensure your place in the workshop.

Where can I find more information about your workshops, including your policies?  ... Go to our website (bioinformatics.ucdavis.edu/training/) and check out our FAQ and Policies.

### Questions

If you have any questions, please don’t hesitate to contact us at training.bioinformatics@ucdavis.edu


### Register

https://registration.genomecenter.ucdavis.edu/events/microbial_community_analysis_Sept_2019/

**Registration is closed for this event**
